src/app/shared/cards/challenge-card/challenge-card.component.ts

```diff
@@ -65,24 +68,44 @@ export class ChallengeCardComponent implements OnInit {
           return of(false);
         }),
         tap((starred: boolean) => (this.starred = starred))
-      );
+      )
+      .subscribe();
   }
 
   toggleStar(event: Event): void {
+    if (this.loggedIn) {
+      this.starred = !this.starred;
+      this.starService().subscribe();
+    } else {
+      this.authService.setRedirectUrl(this.router.url);
+      this.router.navigate(['login']);
+    }
     event.stopPropagation();
-    this.starred = !this.starred;
+  }
+
+  starService(): any {
     if (this.starred) {
-      this.userService.starChallenge(
+      this.numberStarred += 1;
+      return this.userService.starChallenge(
         this.challenge.fullName.split('/')[0],
         this.challenge.name
       );
-      this.numberStarred += 1;
     } else {
-      this.userService.unstarChallenge(
+      this.numberStarred -= 1;
+      return this.userService.unstarChallenge(
         this.challenge.fullName.split('/')[0],
         this.challenge.name
       );
-      this.numberStarred -= 1;
+    }
+  }
+
+  getStarredTooltip(): string {
+    if (this.loggedIn) {
+      return this.starred
+        ? 'Click to remove this from your favorites!'
```

Comment:
@rrchai Do not use exclamation points at the end of messages. I find exclamation point to be "stressful" to the user as exclamation points are supposed to alert and catch the attention of the user. Also try to be consistent will all messages. We already have message without punctuation (e.g. trailing dot) so let's do the same here. Once this is fixed, the PR can be merged.